userspace/libsinsp/metrics_collector.cpp

```diff
@@ -0,0 +1,761 @@
+
+std::string prometheus_metrics_converter::convert_metric_to_text_prometheus(std::string_view metric_name, std::string_view prometheus_namespace, std::string_view prometheus_subsystem, const std::map<std::string, std::string>& const_labels) const
+{
+	return prometheus_exposition_text(prometheus_qualifier(prometheus_namespace, prometheus_subsystem) + std::string(metric_name) + "_info",
+										metric_name,
+										"gauge",
+										"1",
+										const_labels);
+}
+
+void prometheus_metrics_converter::convert_metric_to_unit_convention(metrics_v2& metric) const
+{
+	if((metric.unit == METRIC_VALUE_UNIT_MEMORY_BYTES || metric.unit == METRIC_VALUE_UNIT_MEMORY_KIBIBYTES) &&
+		(metric.type == METRIC_VALUE_TYPE_U32 || metric.type == METRIC_VALUE_TYPE_U64))
+	{
+		if(metric.type == METRIC_VALUE_TYPE_U32)
+		{
+			metric.value.d = libs::metrics::convert_memory(metric.unit, METRIC_VALUE_UNIT_MEMORY_BYTES, metric.value.u32);
+		}
+		else if(metric.type == METRIC_VALUE_TYPE_U64)
+		{
+			metric.value.d = libs::metrics::convert_memory(metric.unit, METRIC_VALUE_UNIT_MEMORY_BYTES, metric.value.u64);
+		}
+		std::string metric_name_str(metric.name);
+		RE2::GlobalReplace(&metric_name_str, s_libs_metrics_units_memory_suffix, "_bytes");
+		strlcpy(metric.name, metric_name_str.c_str(), METRIC_NAME_MAX);
+		metric.type = METRIC_VALUE_TYPE_D;
+		metric.unit = METRIC_VALUE_UNIT_MEMORY_BYTES;
+	}
+	else if(metric.unit == METRIC_VALUE_UNIT_PERC && metric.type == METRIC_VALUE_TYPE_D)
+	{
+		metric.value.d = metric.value.d / 100.0;
+		std::string metric_name_str(metric.name);
+		RE2::GlobalReplace(&metric_name_str, s_libs_metrics_units_perc_suffix, "_ratio");
+		strlcpy(metric.name, metric_name_str.c_str(), METRIC_NAME_MAX);
+		metric.type = METRIC_VALUE_TYPE_D;
+		metric.unit = METRIC_VALUE_UNIT_RATIO;
+	}
+}
+
+void libs_metrics_collector::get_rss_vsz_pss_total_memory_and_open_fds(uint32_t &rss, uint32_t &vsz, uint32_t &pss, uint64_t &host_memory_used, uint64_t &host_open_fds)
+{
+	FILE* f;
+	char filepath[512];
+	char line[512];
+
+	/*
+	 * Get memory usage of the agent itself (referred to as calling process meaning /proc/self/)
+	*/
+
+	f = fopen("/proc/self/status", "r");
+	if(!f)
+	{
+		ASSERT(false);
+		return;
+	}
+
+	while(fgets(line, sizeof(line), f) != NULL)
+	{
+		if(strncmp(line, "VmSize:", 7) == 0)
+		{
+			sscanf(line, "VmSize: %" SCNu32, &vsz);		/* memory size returned in kb */
+		}
+		else if(strncmp(line, "VmRSS:", 6) == 0)
+		{
+			sscanf(line, "VmRSS: %" SCNu32, &rss);		/* memory size returned in kb */
+		}
+	}
+	fclose(f);
+
+	f = fopen("/proc/self/smaps_rollup", "r");
+	if(!f)
+	{
+		ASSERT(false);
+		return;
+	}
+
+	while(fgets(line, sizeof(line), f) != NULL)
+	{
+		if(strncmp(line, "Pss:", 4) == 0)
+		{
+			sscanf(line, "Pss: %" SCNu32, &pss);		/* memory size returned in kb */
+			break;
+		}
+	}
+	fclose(f);
+
+	/*
+	 * Get total host memory usage
+	*/
+
+	snprintf(filepath, sizeof(filepath), "%s/proc/meminfo", scap_get_host_root());
```

Comment:
Question: why are we accessing `proc` here using `scap_get_host_root` but above we are not?
Is it ok?

Reply:
Added comments, I think we needed a clarification, ty!